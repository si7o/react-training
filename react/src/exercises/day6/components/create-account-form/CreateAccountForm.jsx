import React, { useEffect, useReducer, useState } from "react";
import { createAccountFormReducer, RESET, SET_VALUE } from "./reducer";
import Input from "../input/Input";
import Checkbox from "../checkbox/Checkbox";
import Select from "../select/Select";
import styles from "./CreateAccountForm.module.css";
import { useForm } from "./sideEffects";

const initialLoginFormState = {
  name: "",
  lastName: "",
  email: "",
  password: "",
  countryCode: "",
  userAgreement: false,
};

const validationRules = {
  name: "isRequired",
  lastName: "isRequired",
  email: "isEmail",
  password: "isRequired",
  countryCode: "isRequired",
  userAgreement: "isChecked",
};

const validate = (rule, value) => {
  switch (rule) {
    case "isRequired":
      return !value && "This field is required.";
    case "isEmail":
      // TODO
      return !value && "This field is not an Email.";
    case "isChecked":
      return !value && "This field needs to be checked.";
    default:
      return "";
  }
};

const CreateAccountForm = () => {
  const { formState, setValue, getFormData, reset } = useForm(
    initialLoginFormState
  );
  const [errors, setErrors] = useState({});

  const validateInput = (name, value) => {
    const rule = validationRules[name] ?? "";
    const errorMessage = validate(rule, value);

    setErrors({ ...errors, [name]: errorMessage });
  };

  const validateForm = () => {
    const errs = Object.keys(formState).reduce((acc, name) => {
      const rule = validationRules[name] ?? "";
      const value = formState[name].value;
      const errorMessage = validate(rule, value);
      return { ...acc, [name]: errorMessage };
    }, {});
    setErrors(errs);

    return errs;
  };

  const handleFieldChange = (event) => {
    const inputName = event.target.name;
    const inputValue =
      event.target.type === "checkbox"
        ? event.target.checked
        : event.target.value;

    validateInput(inputName, inputValue);

    setValue(inputName, inputValue);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    const errs = validateForm();
    const hasErrors = Object.entries(errs).some(([key, value]) => !!value);
    if (hasErrors) {
      return;
    }

    window.alert(JSON.stringify(getFormData(), null, 2));

    reset({ type: RESET, payload: initialLoginFormState });
  };

  return (
    <div className={styles.createAccountForm}>
      <h3>Create Account</h3>
      <form onSubmit={handleSubmit}>
        <fieldset>
          <Input
            name="name"
            label="First name"
            value={formState.name.value}
            onChange={handleFieldChange}
            errorMessage={errors.name}
          />

          <Input
            name="lastName"
            label="Last name"
            value={formState.lastName.value}
            onChange={handleFieldChange}
            errorMessage={errors.lastName}
          />

          <Input
            name="email"
            label="Email"
            value={formState.email.value}
            onChange={handleFieldChange}
            errorMessage={errors.email}
          />

          <Input
            name="password"
            label="Password"
            type="password"
            value={formState.password.value}
            onChange={handleFieldChange}
            errorMessage={errors.password}
          />

          <Select
            name="countryCode"
            label="Country"
            value={formState.countryCode.value}
            onChange={handleFieldChange}
            errorMessage={errors.countryCode}
            options={[
              { value: "AU", label: "Australia" },
              { value: "RO", label: "Romania" },
              { value: "ES", label: "Spain" },
            ]}
          />
        </fieldset>
        <fieldset>
          <Checkbox
            name="userAgreement"
            label="Accept the terms of service"
            checked={formState.userAgreement.value}
            onChange={handleFieldChange}
            errorMessage={errors.userAgreement}
          />

          <input type="submit" value="Create account" />
        </fieldset>
      </form>
    </div>
  );
};

export default CreateAccountForm;
