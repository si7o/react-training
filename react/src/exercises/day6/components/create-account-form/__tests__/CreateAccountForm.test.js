import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CreateAccountForm from "../CreateAccountForm";

const stubbedExpectedSubmit = {
  name: "Testname",
  lastName: "Test Surname",
  email: "email@test.com",
  password: "Password",
  countryCode: "RO",
  userAgreement: true,
};

const stubbedExpectedKeyboardSubmit = {
  name: "a",
  lastName: "b",
  email: "c",
  password: "d",
  countryCode: "RO",
  userAgreement: true,
};

const mockedAlert = jest.fn();

const jsdomAlert = window.alert;

beforeAll(() => {
  window.alert = mockedAlert;
});

describe("<CreateAccountForm>", () => {
  test("Rendered form matches snapshot", () => {
    const { asFragment } = render(<CreateAccountForm />);
    expect(asFragment).toMatchSnapshot();
  });

  test("Can't submit empty form", () => {
    render(<CreateAccountForm />);

    const submitButton = screen.getByRole("button");
    userEvent.click(submitButton);

    expect(screen.getAllByTestId("errorMessage")).toHaveLength(6);
  });

  test("Can't submit incomplete form", () => {
    render(<CreateAccountForm />);

    const firstField = screen.getAllByRole("textbox")[0];
    userEvent.type(firstField, "Name");

    const submitButton = screen.getByRole("button");
    userEvent.click(submitButton);

    expect(screen.getAllByTestId("errorMessage")).toHaveLength(5);
  });

  test("Can submit form", async () => {
    render(<CreateAccountForm />);

    const nameInput = screen.getByLabelText("First name");
    userEvent.type(nameInput, stubbedExpectedSubmit.name);

    const lastNameInput = screen.getByLabelText("Last name");
    userEvent.type(lastNameInput, stubbedExpectedSubmit.lastName);

    const emailInput = screen.getByLabelText("Email");
    userEvent.type(emailInput, stubbedExpectedSubmit.email);

    const passwordInput = screen.getByLabelText("Password");
    userEvent.type(passwordInput, stubbedExpectedSubmit.password);

    const countrySelect = screen.getByLabelText("Country");
    userEvent.click(countrySelect);
    const selectOptions = screen.getAllByRole("option");
    expect(selectOptions.length).toBeGreaterThan(1);
    userEvent.selectOptions(countrySelect, stubbedExpectedSubmit.countryCode);

    const userAgreement = screen.getByLabelText("Accept the terms of service");
    userEvent.click(userAgreement);

    const submitButton = screen.getByRole("button");
    userEvent.click(submitButton);

    expect(mockedAlert).toHaveBeenCalledWith(
      JSON.stringify(stubbedExpectedSubmit, null, 2)
    );

    const errors = await screen.queryAllByTestId("errorMessage");
    expect(errors).toHaveLength(0);
  });
});

afterAll(() => {
  window.alert = jsdomAlert;
});
