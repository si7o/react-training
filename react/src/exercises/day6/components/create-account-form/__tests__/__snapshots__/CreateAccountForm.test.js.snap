// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateAccountForm> Rendered form matches snapshot 1`] = `[Function]`;
